### DISCUSSION TITLE
Native AOT : Trying to use CompiledBindingExtension instead of ReflectionBindingExtension

### DISCUSSION THREAD

**Discussion by flarive** [AvaloniaUI/Avalonia#13678] [General]
Hello,

I want my Avalonia 11.0.5 app to be compiled with native AOT.
I have a localizer system to localize my resources in multiple languages (Lovcalize.cs and LocalizeExtension.cs)

In LocalizeExtension.cs class i have :

```
var binding = new ReflectionBindingExtension($"[{keyToUse}]")
{
    Mode = BindingMode.OneWay,
    Source = Localize.Instance,
};
return binding.ProvideValue(serviceProvider);
```

Because reflection is not working with native AOT, i have this warning in VS2022 :

Avalonia.Markup.Xaml.MarkupExtensions.ReflectionBindingExtension.ReflectionBindingExtension(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. BindingExpression and ReflectionBinding heavily use reflection. Consider using CompiledBindings instead.

I try to use CompiledBinding instead as suggested by the warning message :

```
var binding = new CompiledBindingExtension()
{
    Mode = BindingMode.OneWay,
    Source = Localize.Instance,
    Path = new CompiledBindingPathBuilder()
};
return binding.ProvideValue(serviceProvider);
```

I don't understand how CompiledBindingPathBuilder class works.
I can't find any way to set the compiled binding path (here i would like to set $"[{keyToUse}]")

Thanks for your help

Eviral

**Comment by maxkatz6** [comment]
Just use observables. Bindings is an XAML concept, there is very little reason to use them in C# code.
If you need IBinding instance, you can also call `observable.ToBinding()`.

See https://learn.microsoft.com/en-us/dotnet/api/system.iobservable-1?view=net-7.0
System.Reactive library might help as well.

**Reply by flarive** [reply]
Hello @maxkatz6,

I really don't understand how I can do that :(
I'm using Sakya Avalonia localization system based upon JSON files.
https://www.sakya.it/wordpress/avalonia-ui-framework-localization/

Unfortunately, Sakya localization system doesn't work in native AOT mode because it uses reflection binding.

I'm trying to update his LocalizeExtension to work in Native AOT mode using CompiledBindingExtension or IObservable<string>... but nothing works :(

Could you please provide me a full Native AOT compliant implementation of LocalizeExtension please ???

Thank you soo much, really stuck on that for several days :(

Eviral

**Reply by maxkatz6** [reply]
You can make the same markup extension with reflection binding AOT friendly:
```
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(Localizer))]
    public override object ProvideValue(IServiceProvider serviceProvider)
```

**Reply by maxkatz6** [reply]
Alternatively, you can skip `LocalizeExtension` completely:
```
Text="{Binding [HelloWorld], Source={x:Static Localizer.Instance}}"
```

It's a bit more verbose but might be better for smaller apps/libraries.

**Reply by flarive** [reply]
Your second solution works perfectly, thanks a lot :)

**Comment by flarive** [comment]
Hum, @maxkatz6, i have another issue :(

I have an app settings dialog in my app with a combobox to let user choose the language he wants.
I would like my dialog to have its localized title updated on the fly.
It's easy to do from XAML but not from code :(


![2024-01-03 17_53_44-Window](https://github.com/AvaloniaUI/Avalonia/assets/7724904/0c062849-297f-44a0-ac36-9c076961abf5)


Here is my dialog creation method (inspired by Material.Avalonia and Avalonia.DialogHost) :

```
private async Task<string> CreateAppSettingsDialog(Window window)
{
    var dialog = CustomDialogHelper.CreateAppSettingsDialog(new DialogWindowBuilderParamsBase
    {
        ContentHeader = Localizer.Localizer.Instance["AppSettings"],
        Width = 520,
        Borderless = true
    });

    DialogResult result = await dialog.ShowDialog(window);

    return result.GetResult;
}
```


ContentHeader is a string in DialogWindowBuilderParamsBase :

```
public class DialogWindowBuilderParamsBase
{
    public string ContentHeader = null;
    ...
}
```

```
public static IDialogWindow<DialogResult> CreateAppSettingsDialog(DialogWindowBuilderParamsBase @params)
{
    var window = new AppSettingsDialog();
    var context = new AppSettingsDialogViewModel(window);

    ApplyBaseParams(context, @params);

    window.DataContext = context;
    SetupWindowParameters(window, @params);
    return new DialogWindowBase<AppSettingsDialog, DialogResult>(window);
}
```



```
public static void ApplyBaseParams<T>(T input, DialogWindowBuilderParamsBase @params)
    where T : DialogWindowViewModel
{
    input.MaxWidth = @params.MaxWidth;
    input.WindowTitle = @params.WindowTitle;
    input.Width = @params.Width;
    input.ContentHeader = @params.ContentHeader;
    input.ContentMessage = @params.SupportingText;
    input.Borderless = @params.Borderless;
    input.WindowStartupLocation = @params.StartupLocation;
    input.DialogShadowKind = (int)@params.ShadowKind;
    ...
}
```


And finally in my dialog view i have :

` <TextBlock Grid.Row="0" Foreground="{DynamicResource MaterialDesignDarkGray}" Padding="0" HorizontalAlignment="Stretch" TextAlignment="Center" VerticalAlignment="Center" FontSize="15" FontWeight="Bold" Text="{CompiledBinding ContentHeader, FallbackValue=''}" />`


ContentHeader is not automatically refreshed when user switch language from the dialog combobox

I would like to use IObservable<string> maybe for ContentHeader but i really don't know how to use it, i would need a minimal simple Avalonia binding to IObservable<string> project.

Thanks a lot again

Eviral

**Comment by flarive** [comment]
Hello again,

I have made a minimal sample project to show localized strings not being refreshed when localized from code, can't deal with that, i'm stuck on that for a few days :(

https://github.com/flarive/AvaloniaLocalizationFromCode

![2024-01-05 08_50_40-Window](https://github.com/AvaloniaUI/Avalonia/assets/7724904/1354f652-f71b-46a2-8520-929c9448a7c6)


Some help would be very great, i'm sure a lot of people using localization in their Avalonia app would be very interested by a solution :)

**Reply by timunie** [reply]
you have a one-time assignment done in your constructor, so it will for sure not update. You can listen to PropertyChanged event here:

```cs
Localizer.Instance.PropertyChanged += (s,e) => MyLocalizedString = Localizer.Instance["AppSettings"];
```

**Reply by flarive** [reply]
Thanks @timunie, going to try :)

It seems nice in this very simple repro case.

But i would need a more generic way, i may have quite a lot of localized properties in my view model that can i need to pass to other viewmodels or classes... Would you have a quite stronger solution that could work for the whole localized app automatically ?

**Reply by timunie** [reply]
```cs
    public string MyLocalizedString => Localizer.Instance["AppSettings"];

    Localizer.Instance.PropertyChanged += (s, e) =>
    {
        this.RaisePropertyChanged(string.Empty);
    };
```

**Reply by flarive** [reply]
Works pretty well, thanks a lot @timunie :)
